---
output: github_document
always_allow_html: yes
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

```{r display image, warning=F,message=F,echo=F}

knitr::include_graphics("C:/Users/samee/OneDrive/Desktop/BOLDConnectR/BOLDconnectR/man/figures/boldconnectr_logo.png")

```



<!-- badges: start -->
<!-- badges: end -->

**BOLDconnectR** is a package designed for **retrieval**, **transformation** and **analysis** of the data available in the *Barcode Of Life Data Systems (BOLD)* database. This package provides the functionality to obtain public and private user data available in the database in the *Barcode Core Data Model (BCDM)* format. Data include information on the **taxonomy**,**geography**,**collection**,**identification** and **DNA barcode sequence** of every submission. The manual is currently hosted here (https://github.com/boldsystems-central/BOLDconnectR_examples/blob/main/BOLDconnectR_1.0.0.pdf)

```{r necessary packages install,include=FALSE}
library(devtools)
library(utils)
library(base)
library(DT)
```

## Installation

The package can be installed using `devtools::install_github` function from the `devtools` package in R (which needs to be installed before installing BOLDConnectR). 
```{r setup,warning=FALSE,message=F}

devtools::install_github("https://github.com/boldsystems-central/BOLDconnectR")

```

```{r BOLDconnectR import,warning=FALSE,message=FALSE}
library(BOLDconnectR)
```

## BOLDconnectR has 10 functions currently:

1. bold.fields.info
2. bold.apikey
3. bold.fetch
4. bold.public.search
5. bold.data.summarize
6. *bold.analyze.align*
7. bold.analyze.tree
8. bold.analyze.diversity
9. bold.analyze.map
10. bold.export

*Function 6 requires the packages `msa` and `Biostrings` to be installed and imported beforehand. Function 7 also uses the the output generated from function 6.*

*NOTE* `msa` and `Biostrings` can be installed using using `BiocManager` package.
```{r msa install,warning=FALSE,message=FALSE}

if (!requireNamespace("BiocManager", quietly=TRUE))
install.packages("BiocManager")

BiocManager::install("msa")
BiocManager::install("Biostrings")

library(msa)
library(Biostrings)

```

### Note on API key

The function `bold.fetch` requires an `api key` internally in order to access and download all public + private user data. API key can be obtained by emailing the BOLD support (support@boldsystems.org). API key can be saved in the R session using `bold.apikey()` function.  

```{r api_key setenv}
# bold_apikey(‘00000000-0000-0000-0000-000000000000’)
```


### A working example of basic BOLDConnectR functionality

API key function must be run prior to using the fetch function (bold.apikey())
```{r apikey,include=F,warning=FALSE,message=F}
bold.apikey('58EEAA88-D681-49AB-A4AC-79779E1AF3D9')
```

```{r fetch data}

BCDM_data<-bold.fetch(get_by = "processid",
                      identifiers = test.data$processid)

knitr::kable(head(BCDM_data,4))

```

Downloaded data can then be summarized in differnt ways 
```{r summarize the data,message=F,warning=F}

BCDM_data_summary<-bold.data.summarize(bold_df = BCDM_data,
                               summarize_by = 'presets',
                               presets = "geography")

BCDM_data_summary$summary

```

Downloaded data can also be exported to the local machine either as a flat file or as a FASTA file for any third party sequence analysis tools.The flat file contents can be modified as per user requirements (entire data or specific presets or individual fields). 

```{r export the data}
# bold.export(bold_df = BCDM_data,
#             export_type = "preset_df",
#             presets = 'taxonomy',
#             export_to = "file_path_with_intended_name")

```


The package also has analyses functions that provide sequence alignment, NJ clustering, biodiversity analysis, occurrence mapping using the downloaded BCDM data. Additionally, these functions also output objects that are commonly used by other R packages ('sf' dataframe, occurrence matrix). 

*BOLDconnectR* is able to fetch public as well as private user data very fast (~100k records in a minute on a fast wired connection) and also offers functionality for data transformation and analysis. 

*Citation:* Padhye SM, Agda TJA, Agda JRA, Ballesteros-Mejia CL, Ratnasingham S. BOLDconnectR: An R package for interacting with the Barcode of Life Data (BOLD) system.(MS in prep)
