---
output: github_document
always_allow_html: yes
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

```{r display image, warning=F,message=F,echo=F,out.width=25}

knitr::include_graphics("C:/Users/samee/OneDrive/Desktop/BOLDConnectR/BOLDconnectR/man/figures/boldconnectr_logo.png")

```



<!-- badges: start -->
<!-- badges: end -->

**BOLDconnectR** is a package designed for **retrieval**, **transformation** and **analysis** of the data available in the *Barcode Of Life Data Systems (BOLD)* database. This package provides the functionality to obtain public and private user data available in the database in the *Barcode Core Data Model (BCDM)* format. Data include information on the **taxonomy**,**geography**,**collection**,**identification** and **DNA barcode sequence** of every submission. The manual is currently hosted here (https://github.com/boldsystems-central/BOLDconnectR_examples/blob/main/BOLDconnectR_1.0.0.pdf)

**BOLDconnectR** requires **R** version **4.0** or above to function properly. The versions of dependent packages have also been set such that they would work with **R >= 4.0**. In addition, there are a few suggested packages that are not mandatory for the package to download and install properly, but, are essential for a couple of functions to work. The names and exact versions of the dependencies/suggestions are given here (https://github.com/boldsystems-central/BOLDconnectR/blob/main/DESCRIPTION). More details on *Suggestions* provided below.

```{r necessary packages install,include=FALSE}
library(devtools)
library(utils)
library(base)
library(DT)
```

## Installation

The package can be installed using `devtools::install_github` function from the `devtools` package in R (which needs to be installed before installing BOLDConnectR). 
```{r setup,warning=FALSE,message=F}

devtools::install_github("https://github.com/boldsystems-central/BOLDconnectR")

```

```{r BOLDconnectR import,warning=FALSE,message=FALSE}
library(BOLDconnectR)
```

## BOLDconnectR has 10 functions currently:

1. bold.fields.info
2. bold.apikey
3. bold.fetch
4. bold.public.search
5. bold.data.summarize
6. *bold.analyze.align*
7. bold.analyze.tree
8. bold.analyze.diversity
9. bold.analyze.map
10. bold.export

*Function 6 requires the packages `msa` and `Biostrings` to be installed and imported beforehand. Function 7 also uses the the output generated from function 6.*

*NOTE* `msa` and `Biostrings` can be installed using using `BiocManager` package.
```{r msa install,warning=FALSE,message=FALSE}

if (!requireNamespace("BiocManager", quietly=TRUE))
install.packages("BiocManager")

BiocManager::install("msa")
BiocManager::install("Biostrings")

library(msa)
library(Biostrings)

```
### Note on API key

The function `bold.fetch` requires an `api key` internally in order to access and download all public + private user data. The API key needed to retrieve BOLD records is found in the BOLD 'Workbench' <https://bench.boldsystems.org/index.php/Login/page?destination=MAS_Management_UserConsole>. After logging in, navigate to 'Your Name' (located at the top left-hand side of the window) and click 'Edit User Preferences'. You can find the API key in the 'User Data' section. Please note that to have an API key available in the workbench, a user must have uploaded at least 10,000 records to BOLD.
API key can be saved in the R session using `bold.apikey()` function.  

```{r api_key setenv}
# Substitute ‘00000000-0000-0000-0000-000000000000’ with your key
# bold.apikey(‘00000000-0000-0000-0000-000000000000’)
```

### Basic usage of BOLDConnectR functionality

API key function must be run prior to using the fetch function (Please see above).
```{r apikey,include=F,warning=FALSE,message=F}
bold.apikey('58EEAA88-D681-49AB-A4AC-79779E1AF3D9')
```

#### Fetch data
```{r fetch data,message=F,warning=F}
BCDM_data<-bold.fetch(get_by = "processid",
                      identifiers = test.data$processid)

knitr::kable(head(BCDM_data,4))
```
Similarly, sampleids or dataset_codes or project_codes can also be used to fetch data. The data can also be filtered on different parameters such as Geography, Attributions and DNA Sequence information using the `_filt` arguments available in the function

#### Summarize downloaded data

Downloaded data can then be summarized in differnt ways. Summaries are generated either on the whole dataset, specific presets (please check the details section of `bold.export()` function in the package manual for details) or specific columns.
```{r summarize the data,message=F,warning=F}
BCDM_data_summary<-bold.data.summarize(bold_df = BCDM_data,
                               summarize_by = 'presets',
                               presets = "geography")

BCDM_data_summary$summary
```
A data summary profile containing aggregates and completeness information is provided as a combination of tabular output and visualization 

#### Export the downloaded data

Downloaded data can also be exported to the local machine either as a flat file or as a FASTA file for any third party sequence analysis tools.The flat file contents can be modified as per user requirements (entire data or specific presets or individual fields). 

```{r export the data}
# Preset dataframe
# bold.export(bold_df = BCDM_data,
#             export_type = "preset_df",
#             presets = 'taxonomy',
#             export_to = "file_path_with_intended_name")

# Unaligned fasta file
# bold.export(bold_df = BCDM_data,
#             export_type = "fas",
#             cols_for_fas_names = c("bin_uri","genus","species"),
#             export_to = "file_path_with_intended_name")

```

#### Other functions

The package also has analyses functions that provide sequence alignment, NJ clustering, biodiversity analysis, occurrence mapping using the downloaded BCDM data. Additionally, these functions also output objects that are commonly used by other R packages ('sf' dataframe, occurrence matrix). Please go through the help manual (Link provided above) for detailed usage of all the functions of BOLDConnectR with examples.

*BOLDconnectR* is able to retrieve data very fast (~100k records in a minute on a fast wired connection).

*Citation:* Padhye SM, Agda TJA, Agda JRA, Ballesteros-Mejia CL, Ratnasingham S. BOLDconnectR: An R package for interacting with the Barcode of Life Data (BOLD) system.(MS in prep)
