#'Export files generated by BOLDconnectR
#'
#' @description
#'The function is used to export some of the output data generated by BOLDconnectR
#'
#' @param bold_df The data.frame either retrieved from [bold.fetch()],`bold.analyze.align` or a user modified BCDM dataset.
#' @param export_type A character input specifying the type of output required. Should be either of "preset_df","msa" or "fas".
#' @param presets A single character vector specifying a preset for which a data summary is sought (Check the `details` section for more information). Default value is NULL.
#' @param cols_for_fas_names A single or multiple character vector indicating the column headers that should be used to name each sequence for the unaligned FASTA file. Default is NULL; in this case, only the processid is used as the name.
#' @param export_to A character value specifying the data path and the name for the file. Extension should not be provided
#
#' @details
#' `bold.export` offers an export option for some of the sequence-based outputs obtained from functions within the `BOLDconnectR` package as well as a `preset` defined modified BCDM dataframe. Sequence information downloaded using [bold.fetch()] or the aligned sequences obtained using `bold.analyze.align` can be exported as a FASTA file for any third party tool (via `export_type`=’fas’ or ’msa’). Data fetched by [bold.fetch()] can be directly used to export the unaligned FASTA file, while the modified dataframe from `bold.analyze.align` is required for exporting the multiple sequence alignment. `presets`are defined set of columns representing a specific aspect of the BOLD BCDM data (currently available are: `taxonomy`,`geography`,`sequences`,`attributions`, `ecology_biogeography` and `other_meta_data`)
#'The name for individual sequences in the unaligned FASTA file output can be customized by using the `cols_for_fas_names` argument. If more than one field is specified, the name will follow the sequence of the fields given in the vector. The multiple sequence aligned FASTA file uses the same name provided by the user in the `bold.analyze.align` function. Additionally, this function allows for the export of user-edited data (in taxonomy, geography etc.) as a csv/tsv file while retaining its BCDM format. This functionality is developed with the future potential of uploading data to BOLD using the package. Edits to the BCDM data can be made using any other R packages so long as it maintains the BCDM format.
#'
#' @examples
#' \dontrun{
#' data_for_export_ids <- bold.public.search(taxonomy = "Poecilia reticulata",
#'                                           filt_basecount= c(500,600),
#'                                           filt_marker = "COI-5P")
#' # Fetching the data using the ids
#' data_for_export <- bold.fetch(get_by = "processid",
#'                               identifiers = data_for_export_ids$processid)
#'
#' # Export the BCDM data using 'presets' (Using getwd() as the path and trial_export as the name)
#'
#' taxonomy_df<-bold.export(bold_df=data_for_export,
#'                          export_type = "preset_df",
#'                          presets = taxonomy,
#'                          export_to = paste(getwd(),'/','trial_export',sep=""))
#'
#'
#' # Align the data (using processid and bin_uri as fields for sequence names)
#' # Users need to install and load packages `msa` and `Biostrings` before using bold.analyze.align.
#'
#' seq_align<-bold.analyze.align(data_for_export,
#'                               cols_for_seq_names = c("processid","bin_uri"),
#'                               align_method = "ClustalOmega")
#' # Export the multiple sequence alignment
#' # Note the input data here is the modified BCDM data after using bold.analyze.align.
#' # file.path and file.name must be provided
#'
#' # Export the fasta file (unaligned)
#' # Note that input data is the original BCDM data retrieved
#' using bold.fetch (Using getwd() as the path and trial_export as the name).
#' bold.export(bold_df = data_for_export,
#'             export_type = "fas",
#'             cols_for_fas_names = c("bin_uri","genus","species"),
#'             export_to = paste(getwd(),'/','trial_export',sep=""))
#' }
#'
#' @returns It exports a .fas or a csv/tsv file based on the export argument.
#'
#' @importFrom utils write.table
#' @importFrom ape read.dna
#' @importFrom ape write.FASTA
#'
#' @export
#'
bold.export<-function(bold_df,
                      export_type=c("preset_df","msa","fas"),
                      presets=NULL,
                      cols_for_fas_names=NULL,
                      export_to)

{
  # Check if data is a data frame object

  if (!is.data.frame(bold_df)) {
    stop("Input data should be a BCDM dataframe")
  }


  # Check whether the data frame is empty

  if (nrow(bold_df)==0) {
    stop("Input data should be a BCDM dataframe")
  }

  switch(export_type,

         "preset_df" =

           {

             if (is.null(presets))
             {
               stop("One of the presets must be provided when export_type = preset_df")
             }


             preset_data=check_and_return_preset_df(bold_df,
                                                    category = "check_return",
                                                    preset = presets)

             utils::write.table(preset_data,
                                paste0(export_to,".tsv",sep=""),
                                sep = "\t",
                                row.names = FALSE,
                                quote = FALSE)

           },

         "msa" =

           {

             if (any(!is.null(cols_for_fas_names),!is.null(presets)))
             {
               stop("Please remove any presets or field names provided in the presets or 'cols_for_fas_names' arguments.")
             }

             seq.data=bold_df%>%
               dplyr::filter(!is.na(nuc))%>%
               dplyr::filter(!is.null(nuc))%>%
               dplyr::mutate(nuc=gsub("-","",nuc))%>%
               dplyr::filter(nuc!="")%>%
               dplyr::select(matches("^aligned_seq$",ignore.case=TRUE),
                             matches("^msa.seq.name$",ignore.case=TRUE))%>%
               dplyr::rename('seq.name'='msa.seq.name')



             ## Export the result as a fasta file.

             result=generate_ape_file(data = seq.data,
                                      align_unaligned = "aligned_seq")


             ape::write.FASTA(result,
                              file=paste0(export_to,".fas",sep=""))

           },

         "fas" =

           {
             if (is.null(presets)==FALSE)
             {
               stop("Please remove any presets provided in the 'presets' arguments.")
             }

             seq.data=bold_df%>%
               dplyr::select(matches("^nuc$",ignore.case=TRUE),
                             all_of(cols_for_fas_names))%>%
               dplyr::filter(!is.na(nuc))%>%
               dplyr::filter(!is.null(nuc))%>%
               dplyr::mutate(nuc=gsub("-","",nuc))%>%
               dplyr::filter(nuc!="")%>%
               dplyr::rowwise()%>%
               dplyr::mutate(across(all_of(cols_for_fas_names),
                                    as.character))%>%
               dplyr::select(nuc,
                             all_of(cols_for_fas_names))%>%
               dplyr::mutate(seq.name=paste0(paste(as.character(c_across(all_of(cols_for_fas_names))),
                                                   collapse = "|")))%>%
               dplyr::ungroup()


             # Export the result as a raw fasta file.

             result=generate_ape_file(data = seq.data,
                                      align_unaligned = "nuc")


             ape::write.FASTA(result,
                              file=paste0(export_to,".fas",sep=""))

           }

  )


}
